### PR TITLE
Feature/validate with language and refactor

### PR DESCRIPTION
**Pull request by joswarmer**
Yat another huge PR, will try to make more focussed ones in the future.

### CODE REVIEW

File: packages/validation/src/validators/LionWebSyntaxValidator.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ export class LionWebSyntaxValidator {
     /**
      * Check whether `obj` is a JSON object that conforms to the serialization syntax of LionCore.
      * All errors found will be pushed into the `errors` array, if its length is not 0, the check has failed.
-     * @param obj
+     * @param obj
```
Comment (dslmeinte) [inline packages/validation/src/validators/LionWebSyntaxValidator.ts:65]: Ooh, you're **so** triggering my OCD! :D

File: packages/validation/src/validators/LionWebLanguageValidator.ts [hunk 2]
```diff
@@ -1,39 +1,84 @@
...
+            const propClassifier = this.registry.getNodeByMetaPointer(actualProp.property)
+            if (propClassifier === undefined) {
+                this.validationResult.issue(
+                    new GenericIssue(new JsonContext(null, ["nodes", "properties"]), 
+                        `Property ${actualProp.property.key} not found for classifier ${classifier.id}`)
+                )
+            }
+            // this.availableLanguages.getClassifierProperty(classifier, propClassifier)
+        })
+        // node.containments
```
Comment (dslmeinte) [inline packages/validation/src/validators/LionWebLanguageValidator.ts:80]: Do these comments mean they still need to be implemented?

**Review by dslmeinte — APPROVED** [review]
Quite difficult to thoroughly review such a big PR. Maybe we should do more pair programming?

File: packages/validation/src/util/graphs.ts
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/validation/src/util/graphs.ts:1]: Indeed, that is where I got it from :-)

File: packages/validation/src/validators/LionWebLanguageValidator.ts (hunk 2 above)
Comment (joswarmer) [reply packages/validation/src/validators/LionWebLanguageValidator.ts:80]: No, will remove

File: packages/validation/src/languages/MetaPointerMap.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+export class MetaPointerMap {
+    /**
+     * Map from languageKey => languageVersion => key => LionWebJsonNode
+     */
+    map: Map<string, Map<string, Map<string, LionWebJsonNode>>> = new Map<string, Map<string, Map<string, LionWebJsonNode>>>()
+    
+    add(languageKey: string, languageVersion: string, node: LionWebJsonNode): void {
+        const keyProperty = NodeUtils.findProperty(node, MetaPointers.IKeyedKey)
+        if (keyProperty === undefined) {
+            console.log("MetaPointerMap.add: trying to add node without keyn property.")
```
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/validation/src/languages/MetaPointerMap.ts:12]: Will fix

File: packages/validation/src/json/M3definitions.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+export const LION_CORE_M3_NAME = "LionCore_M3"
+
+export const LION_CORE_M3_VERSION = "2023.1"
+// Builtin datatypes
+export const LIONWEB_BOOLEAN_TYPE = "LionCore-builtins-Boolean"
+export const LIONWEB_JSON_TYPE = "LionCore-builtins-JSON"
+export const LIONWEB_INTEGER_TYPE = "LionCore-builtins-Integer"
+export const LIONWEB_STRING_TYPE = "LionCore-builtins-String"
+
+// TODO Trial: remove when not used.
```
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/validation/src/json/M3definitions.ts:17]: yes, will do

File: packages/validation/src/json/LionWebJsonChunkWrapper.ts [hunk 5]
```diff
@@ -1,36 +1,52 @@
...
+    constructor(chunk: LionWebJsonChunk) {
+        this.jsonChunk = chunk
+        this.prepareNodeIds(chunk)
     }
 
+    /**
+     * Merge `chunk` into `this.jsonChunk`
+     * @param addedChunk
+     */
+    // pushChunk(addedChunk: LionWebJsonChunk) {
```
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/validation/src/json/LionWebJsonChunkWrapper.ts:39]: no, will remove it

File: .editorconfig [hunk 6]
```diff
@@ -8,7 +8,9 @@ indent_style = space
 indent_size = 4
 
 [*.json]
-indent_size = 2
+indent_size = 4
```
[the remark this answers is not in the document]
Comment (joswarmer) [reply .editorconfig:11]: Looking at the code, indent size currently is 4 for most (or all, I did not check) typescript files.  So I just fixed this setting to be equal to the indent size that we are already using. Would you want to change this to 2?

File: packages/cli/src/validator.ts [hunk 7]
```diff
@@ -1,9 +1,9 @@
-import {validateFileResultWithLanguage} from "@lionweb/validation"
+import { LanguageRegistry, validateFileResult } from "@lionweb/validation"
```
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/cli/src/validator.ts:1]: I use prettier and that results in the extra space, so these are the current prettier settings. I checked and see that we both seem to use different style here.

File: packages/utilities/src/serialization/ordering.ts [hunk 8]
```diff
@@ -12,8 +12,8 @@ import {
 
 const orderedMetaPointer = ({language, version, key}: MetaPointer): MetaPointer =>
     ({
-        language,
-        version,
+        language: language,
```
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/utilities/src/serialization/ordering.ts:15]: I don't recall that I changed this file, so I changed this back.

File: packages/validation/src/util/graphs.ts
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/validation/src/util/graphs.ts:1]: Added a comment

File: packages/test/testlanguage/LionCore_M3.json
[the remark this answers is not in the document]
Comment (joswarmer) [reply packages/test/testlanguage/LionCore_M3.json:1]: Leave this for now, need an issue to cleanup these jsons.

File: packages/validation/src/validators/LionWebSyntaxValidator.ts (hunk 1 above)
Comment (joswarmer) [reply packages/validation/src/validators/LionWebSyntaxValidator.ts:65]: ?

File: packages/validation/src/validators/LionWebLanguageValidator.ts (hunk 2 above)
Comment (joswarmer) [reply packages/validation/src/validators/LionWebLanguageValidator.ts:80]: Actually the validation is not complete yet, but these comment are out of date.

**Pull request merged**